Sources/Utility/FSWatch.swift

```diff
@@ -0,0 +1,602 @@
+        case 12: return set.__fds_bits.12 & mask != 0
+        case 13: return set.__fds_bits.13 & mask != 0
+        case 14: return set.__fds_bits.14 & mask != 0
+        case 15: return set.__fds_bits.15 & mask != 0
+        default: return false
+    }
+}
+
+#endif
+
+// MARK:- FSEventStream
+
+#if canImport(Darwin)
+
+private func callback(
+    streamRef: ConstFSEventStreamRef,
+    clientCallBackInfo: UnsafeMutableRawPointer?,
+    numEvents: Int,
+    eventPaths: UnsafeMutableRawPointer,
+    eventFlags: UnsafePointer<FSEventStreamEventFlags>,
+    eventIds: UnsafePointer<FSEventStreamEventId>
+) {
+    let eventStream = unsafeBitCast(clientCallBackInfo, to: FSEventStream.self)
+
+    // We expect the paths to be reported in an NSArray because we requested CFTypes.
+    let eventPaths = unsafeBitCast(eventPaths, to: NSArray.self) as? [String] ?? []
+
+    // Compute the set of paths that were changed.
+    let paths = eventPaths.map({ AbsolutePath($0) })
+
+    eventStream.callbacksQueue.async {
+        eventStream.delegate.pathsDidReceiveEvent(paths)
+    }
+}
+
+public protocol FSEventStreamDelegate {
+    func pathsDidReceiveEvent(_ paths: [AbsolutePath])
+}
+
+/// Wrapper for Darwin's FSEventStream API.
+public final class FSEventStream {
+
+    /// The errors encountered during fs event watching.
+    public enum Error: Swift.Error {
+        case unknownError
+    }
+
+    /// Reference to the underlying event stream.
+    ///
+    /// This is var and implicitly unwrapped optional because
+    /// we need to capture self for the context.
+    private var stream: FSEventStreamRef!
+
+    /// Reference to the handler that should be called.
+    let delegate: FSEventStreamDelegate
+
+    /// The thread on which the stream is running.
+    private var thread: Basic.Thread?
+
+    /// The run loop attached to the stream.
+    private var runLoop: CFRunLoop?
+
+    /// Callback queue for the delegate.
+    fileprivate let callbacksQueue = DispatchQueue(label: "org.swift.swiftpm.\(FSEventStream.self).callback)")
+
+    public init(
+        paths: [AbsolutePath],
+        delegate: FSEventStreamDelegate
+    ) {
+        self.delegate = delegate
+
+        // Create the context that needs to be passed to the callback.
+        var callbackContext = FSEventStreamContext()
+        callbackContext.info = unsafeBitCast(self, to: UnsafeMutableRawPointer.self)
+
+        // Create the stream.
+        self.stream = FSEventStreamCreate(nil,
+            callback, 
+            &callbackContext, 
+            paths.map({ $0.asString }) as CFArray, 
+            FSEventStreamEventId(kFSEventStreamEventIdSinceNow),
+            /* latency */ 1,
+            // FIXME: This needs to be customizable.
+            FSEventStreamCreateFlags(kFSEventStreamCreateFlagUseCFTypes)
+        )
+    }
+
+    // Start the runloop.
+    public func start() throws {
+        let thread = Thread { [weak self] in
+            guard let `self` = self else { return }
```

Comment:
Is setting up a thread loop really necessary ? FSEventStream can be setup to pass events to a dispatch queue via `FSEventStreamSetDispatchQueue`

Reply:
ah, good idea. I didn't know about that option